palace/utils/geodata.cpp

```diff
@@ -89,55 +103,114 @@ std::unique_ptr<mfem::ParMesh> ReadMesh(MPI_Comm comm, const IoData &iodata, boo
+                    refinement.nonconformal,
                 "If there are wedge elements, AMR must be nonconformal!");
+    MFEM_VERIFY(!use_amr || iodata.model.make_simplex || !element_types.has_pyramids ||
+                    refinement.nonconformal,
+                "If there are pyramid elements, AMR must be nonconformal!");
+
+    // Optionally convert mesh elements to simplices, for example in order to enable
+    // conformal mesh refinement, or hexes.
+    if (iodata.model.make_simplex || iodata.model.make_hex)
+    {
+      SplitMeshElements(smesh, iodata.model.make_simplex, iodata.model.make_hex);
+    }
 
     // Optionally reorder elements (and vertices) based on spatial location after loading
     // the serial mesh.
-    if (reorder)
+    if (iodata.model.reorder_elements)
     {
-      ReorderMesh(*smesh);
+      ReorderMeshElements(*smesh);
+    }
+
+    // Refine the serial mesh (not typically used, prefer parallel uniform refinement
+    // instead).
+    {
+      int ne = smesh->GetNE();
+      for (int l = 0; l < iodata.model.refinement.ser_uniform_ref_levels; l++)
+      {
+        smesh->UniformRefinement();
+      }
+      if (iodata.model.refinement.ser_uniform_ref_levels > 0)
+      {
+        Mpi::Print("Serial uniform mesh refinement levels added {:d} elements (initial = "
+                   "{:d}, final = {:d})\n",
+                   smesh->GetNE() - ne, ne, smesh->GetNE());
+      }
     }
 
     // Clean up unused domain elements from the mesh, add new boundary elements for material
     // interfaces if not present. Can only clean up conforming meshes, assumes that any
     // nonconformal mesh was generated by adaptation and thus does not need checking.
-    if ((clean_elem || add_bdr) && smesh->Conforming())
+    if (smesh->Conforming())
     {
-      static_cast<void>(CheckMesh(smesh, iodata, clean_elem, add_bdr, false, nullptr));
+      static_cast<void>(CheckMesh(smesh, iodata, iodata.model.clean_unused_elements,
+                                  iodata.model.add_bdr_elements));
     }
 
     // Generate the mesh partitioning.
-    partitioning = GetMeshPartitioning(*smesh, Mpi::Size(comm), iodata.model.partition);
-
-    // Optionally add subdomain interface boundary elements.
-    if (add_subdomain && smesh->Conforming())
-    {
-      static_cast<void>(
-          CheckMesh(smesh, iodata, false, false, add_subdomain, partitioning.get()));
-    }
+    partitioning = GetMeshPartitioning(*smesh, Mpi::Size(comm), iodata.model.partitioning);
   }
 
+  // Distribute the mesh.
   std::unique_ptr<mfem::ParMesh> pmesh;
   if (use_mesh_partitioner)
   {
     pmesh = DistributeMesh(comm, smesh, partitioning.get(), iodata.problem.output);
   }
   else
   {
+    // Send the preprocessed serial mesh and partitioning as a byte string.
+    constexpr bool generate_edges = false, refine = true, fix_orientation = false;
+    std::string so;
+    int slen = 0;
```

Comment:
I think this implicitly is setting a limit of 2.147483647 GB for any mesh (after modifications). I'm not sure if this is a limit that might be practically hit, it seems pretty high. Could consider using `std::size_t` directly, as that would be 4.28. However, the catch between `int` and `std::size_t` for the overflow ought to warn if the size is between 2.14 and 4.28 though, so we'd get a good amount of warning. And if the byte string is going past 4.28 GB then this whole approach would need reconsidering anyway.

Reply:
The problem is that the MPI interface only allows for `int` message lengths. We have an assertion to check for the overflow but indeed for a big enough mesh we would/will need to segment the message as multiple broadcast messages. Also just an aside but isnt `std::size_t` unsigned 64 bit on a 64 bit machine? If so the maximum value is well beyond 4.28e9.

Reply:
You're right, I had just been thinking of `std::size_t` as a synonym for unsigned int, but you're right it's vastly larger, so that overflow check should be absolutely plenty.